### PR TITLE
implement findNMostRecentTransactions with tests

### PR DESCRIPTION
**Pull request by alon-e**
#88 
only the logic and tests, missing integration.

### CODE REVIEW

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.model.Hash;
+import com.iota.iri.storage.Tangle;
+
+import java.util.*;
+
+public class ConnectedComponentsCalculator {
+    public final Tangle tangle;
+
+    public static final int N = 5000;
```
Comment (alongalky) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java]: More descriptive constant name. Maybe `MAX_TRANSACTIONS`?

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
```
Comment (alongalky) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java:13]: Recommend refactoring `N` to be a parameter for either the constructor or `findNMostRecentTransactions`, for two reasons:
1. Testing. No need to create tangles of size 5K in each test, and also you can test behavior different values of N if relevant.
1. Avoid proliferation of magic numbers

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+            if (!result.contains(current.getBranchTransactionHash())
+            && !current.getTrunkTransactionHash().equals(current.getBranchTransactionHash())) {
+                queue.add(current.getBranchTransaction(tangle));
+            }
+        }
+
+        return result;
+    }
+
+    //find connected components in `N` (step 2)
```
Comment (alongalky) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java]: Remove comments, or at least add a TODO: prefix

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+public class ConnectedComponentsCalculator {
+    public final Tangle tangle;
+
+    public static final int N = 5000;
+
+    public ConnectedComponentsCalculator(Tangle tangle) {
+        this.tangle = tangle;
+    }
+
+    //find N most recent transactions
```
Comment (alongalky) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java]: Recommend removing this comment, it doesn't add anything

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        }
+
+        return result;
+    }
+
+    //find connected components in `N` (step 2)
+
+    //randomly select a tip from the largest CC (step 3)
+
+    private TransactionViewModel fromHash(Hash hash) {
```
Comment (alongalky) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java:49]: I guess it's a matter of taste, but IMO it's not worth it to add this code just to gain the `stream()` ability. You also lose exception handling. There isn't a test for the case that one of the hashes is null (which is what happens if `fromHash` fails), and I'm not sure what would happen. It might throw an exception afterwards, or maybe think it's the genesis?

Comment (alongalky) [reply src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java:49]: I recommend switching to a loop, but if you want to keep the stream you need a null check before adding to the queue

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            TransactionViewModel current = queue.poll();
+            result.add(current.getHash());
+
+            //add children to priority queue, only if not already chosen.
+            if (!result.contains(current.getTrunkTransactionHash())) {
+                queue.add(current.getTrunkTransaction(tangle));
+            }
+            if (!result.contains(current.getBranchTransactionHash())
+            && !current.getTrunkTransactionHash().equals(current.getBranchTransactionHash())) {
+                queue.add(current.getBranchTransaction(tangle));
```
Comment (alongalky) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java:42]: OCD optimization: I think it might be better to switch the operands in the && expression

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+        //the heap is initialized with the tips.
+        tips.stream().map(this::fromHash).forEach(queue::add);
+
+        while (!queue.isEmpty() && result.size() < N) {
+            TransactionViewModel current = queue.poll();
+            result.add(current.getHash());
+
+            //add children to priority queue, only if not already chosen.
+            if (!result.contains(current.getTrunkTransactionHash())) {
```
Comment (alongalky) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java]: The if statements may be more readable if you define `trunk` and `branch` variables.

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+public class ConnectedComponentsCalculatorTest {
+
+    private TemporaryFolder dbFolder;
+    private TemporaryFolder logFolder;
+    private Tangle tangle;
+    private TipsViewModel tipsViewModel;
+
+    @Rule
+    public ExpectedException exception = ExpectedException.none();
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: Is this rule used?

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        dbFolder = new TemporaryFolder();
+        logFolder = new TemporaryFolder();
+        tangle = new Tangle();
+        dbFolder.create();
+        logFolder.create();
+        tangle.addPersistenceProvider(new RocksDBPersistenceProvider(dbFolder.getRoot().getAbsolutePath(),
+                logFolder.getRoot().getAbsolutePath(), 1000));
+        tangle.init();
+
+        tipsViewModel = Mockito.mock(TipsViewModel.class);
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: I think you don't need to mock tipsViewModel, just send the tip set directly. This would only be required if the constructor received a tipsViewModel and the call to `getRandomSolidTipHash` happened inside the class

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    //hairy genesis
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: Is this a TODO? Let's remove these comments anyway, the tests are either important enough to code or not

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+
+        Mockito.when(tipsViewModel.getRandomSolidTipHash()).thenReturn(transaction.getHash());
+
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                Collections.singleton(tipsViewModel.getRandomSolidTipHash()));
+
+        Assert.assertTrue(recentTransactions.contains(transaction.getHash()));
+        Assert.assertTrue(recentTransactions.contains(Hash.NULL_HASH));
+
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: stray newline

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    @Test
+    public void returnsGenesisTxInChain() throws Exception {
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java:64]: You can copy paste makeChain from here, since you have 3 tests that use one: https://github.com/iotaledger/cliri/blob/483be1c6bf019b1e7a33e10e091141148557c39e/src/test/java/com/iota/iri/service/tipselection/impl/EntryPointSelectorCumulativeWeightThresholdTest.java#L212

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+            transaction.setArrivalTime(incr++);
+            transaction.store(tangle);
+        }
+
+        Mockito.when(tipsViewModel.getRandomSolidTipHash()).thenReturn(transactions.get(transactions.size() - 1).getHash());
+
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                Collections.singleton(tipsViewModel.getRandomSolidTipHash()));
+        Assert.assertTrue(recentTransactions.contains(Hash.NULL_HASH));
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java:71]: Maybe also `Assert.assertEquals(chainLength + 1, recentTransactions.size())` ?

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+            transaction.setArrivalTime(incr++);
+            transaction.store(tangle);
+        }
+
+        Mockito.when(tipsViewModel.getRandomSolidTipHash()).thenReturn(transactions.get(transactions.size() - 1).getHash());
+
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                Collections.singleton(tipsViewModel.getRandomSolidTipHash()));
+        Assert.assertFalse(recentTransactions.contains(Hash.NULL_HASH));
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java:82]: Maybe also `Assert.assertEquals(ConnectedComponentsCalculator.N, recentTransactions.size())` ?

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+        }
+
+        Collection<Hash> transactionHashes = transactions.stream().map(TransactionViewModel::getHash).collect(Collectors.toList());
+
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                transactionHashes);
+        Assert.assertTrue(recentTransactions.containsAll(transactionHashes));
+        Assert.assertTrue(recentTransactions.contains(Hash.NULL_HASH));
+
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: newline

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+        Mockito.when(tipsViewModel.getRandomSolidTipHash()).thenReturn(transactions.get(transactions.size() - 1).getHash());
+
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                Collections.singleton(tipsViewModel.getRandomSolidTipHash()));
+        Assert.assertFalse(recentTransactions.contains(Hash.NULL_HASH));
+    }
+
+    @Test
+    public void starAroundGenesis() throws Exception {
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: change name to reflect what's being tested (`allTipsAndGenesisReturnedForStarAroundGenesis`)

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+        }
+
+        Collection<Hash> transactionHashes = transactions.stream().map(TransactionViewModel::getHash).collect(Collectors.toList());
+
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                transactionHashes);
+        Assert.assertFalse(recentTransactions.contains(Hash.NULL_HASH));
+        Assert.assertFalse(recentTransactions.containsAll(transactionHashes)); // there exists one tip not in list
+
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: newline

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                transactionHashes);
+
+        Assert.assertTrue(recentTransactions.containsAll(
+                transactionChain.stream().map(TransactionViewModel::getHash).collect(Collectors.toList()))); // there exists one tip not in list
+        Assert.assertFalse(recentTransactions.containsAll(transactionHashes)); // there exists one tip not in list
+        Assert.assertFalse(recentTransactions.contains(Hash.NULL_HASH));
+
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java:155]: newline

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+            transaction.store(tangle);
+        }
+
+        Collection<Hash> transactionHashes = transactions.stream().map(TransactionViewModel::getHash).collect(Collectors.toList());
+
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                transactionHashes);
+        Assert.assertFalse(recentTransactions.contains(Hash.NULL_HASH));
+        Assert.assertFalse(recentTransactions.containsAll(transactionHashes)); // there exists one tip not in list
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: Instead of this assertion (or maybe in addition), I'd put the more specific assertion that the oldest tx is the one that's excluded.

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                transactionHashes);
+        Assert.assertTrue(recentTransactions.containsAll(transactionHashes));
+        Assert.assertTrue(recentTransactions.contains(Hash.NULL_HASH));
+
+    }
+
+    @Test
+    public void starAroundGenesisTooBig() throws Exception {
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: descriptive name (`oldTipsDroppedWhenStarAroundGenesisIsTooBig`)

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 21]
```diff
@@ -0,0 +1,240 @@
...
+
+        List<Hash> transactionHashes = transactions.stream().map(TransactionViewModel::getHash).collect(Collectors.toList());
+        transactionHashes.add(transactionChain.get(transactionChain.size() - 1).getHash());
+
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                transactionHashes);
+
+        Assert.assertTrue(recentTransactions.containsAll(
+                transactionChain.stream().map(TransactionViewModel::getHash).collect(Collectors.toList()))); // there exists one tip not in list
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: I think the comment at the end of the line is incorrect

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java (hunk 12 above)
Comment (alongalky) [reply src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java:64]: Maybe also write makeStarAroundGenesis (to make the tests shorter)

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+
+    }
+
+    @Test
+    public void starAroundGenesisAndChain() throws Exception {
+        long incr = 1000;
+
+        final int amount = ConnectedComponentsCalculator.N / 10;
+        //star
+        List<TransactionViewModel> transactions = new ArrayList<TransactionViewModel>();
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: Rename to `transactionsStar`

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+            transactionChain.add(new TransactionViewModel(
+                    getRandomTransactionWithTrunkAndBranch(prevTxHash, prevTxHash), getRandomTransactionHash()));
+        }
+
+        for (TransactionViewModel transaction : transactionChain) {
+            transaction.setArrivalTime(incr++);
+            transaction.store(tangle);
+        }
+
+        List<Hash> transactionHashes = transactions.stream().map(TransactionViewModel::getHash).collect(Collectors.toList());
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: `starTransactionHashes`

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+        ConnectedComponentsCalculator connectedComponentsCalculator = new ConnectedComponentsCalculator(tangle);
+        Collection<Hash> recentTransactions = connectedComponentsCalculator.findNMostRecentTransactions(
+                transactionHashes);
+        Assert.assertFalse(recentTransactions.contains(Hash.NULL_HASH));
+        Assert.assertFalse(recentTransactions.containsAll(transactionHashes)); // there exists one tip not in list
+
+    }
+
+    @Test
+    public void starAroundGenesisAndChain() throws Exception {
```
Comment (alongalky) [inline src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: This is a great test but a bit hard to follow. Recommend putting the constants at the top (i.e. `chainLength`) and being more explicit in the test name `mostRecentTipsChosenFromStarAroundGenesisAndChain` (probably can be improved)

**Review by alongalky — CHANGES_REQUESTED** [review]

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java (hunk 6 above)
Comment (alon-e) [reply src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java:42]: I think this is more readable (the addition of equality check doesn't convey much logic) -- I can reorg it to be cleaner.

File: src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java (hunk 8 above)
Comment (alon-e) [reply src/test/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculatorTest.java]: no, thx (copy paste template)

File: src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java [hunk 25]
```diff
@@ -9,51 +9,48 @@
 public class ConnectedComponentsCalculator {
     public final Tangle tangle;
 
-    public static final int N = 5000;
+    private final int maxTransaction;
```
Comment (alongalky) [inline src/main/java/com/iota/iri/service/tipselection/impl/ConnectedComponentsCalculator.java]: `maxTransactions` :)

**Review by alongalky — APPROVED** [review]
